### ISSUE TITLE
Could we bundle Protocol CSS more efficiently?

### ISSUE THREAD

**Issue by alexgibson** [mozilla/bedrock#11032]
## Description

We currently do over 657 CSS `@imports` for Protocol components in over 187 different .scss files. These files are then compiled to over 161 CSS bundles that are used across different pages on bedrock.

This is **a lot** of CSS, especially given that we're now using a component-based design system across the site. There will always be snowflake pages on mozorg and I think that's OK, but I am curious if we could be compiling our CSS in a better way? 

For example, instead of `@import`'ing each component that a page needs into `page.scss`, might it be better for pages to use common pre-baked bundles of components (e.g. `c-hero.css`, `c-split.css`), in a similar way to Contentful pages? This would add a few extra HTTP requests to pages, but that shouldn't be a huge performance impact now were on HTTP/2. It would also help improve load time between page navigations, as right now the same chunks of CSS are re-downloaded each time.

This is just an idea for thought / discussion. I'd be interested in thoughts from others here.

**Comment by nathan-barrett** [comment]
Thanks for putting this together @alexgibson! In your example would we create a "pre-baked" bundle for each of the protocol components? Would we then add that to the static-bundles.json for that specific bundle for a page?

Also wondering if we consistently import the protocol/lib file into each of our css files, ( I feel like I've seen it on every file i've worked with). I wonder if we could just import it into the protocol-base file we use.

**Comment by maureenlholland** [comment]
@nathan-barrett for the lib import, I think we need it to access the sass variables and functions that get compiled away in the base stylesheets before the specific components can use them

I definitely think there's room to be more efficient with Protocol though. The component css bundles sound like a good place to start.

**Comment by alexgibson** [comment]
> In your example would we create a "pre-baked" bundle for each of the protocol components? 

At least for components we use most commonly across the site, yes.

> Would we then add that to the static-bundles.json for that specific bundle for a page?

No we would just create a bundle for the components, and then page templates could use that directly without needing to bundle things again: e.e. `{{ css_bundle('c-split') }}`

**Comment by nathan-barrett** [comment]
@maureenlholland ah ok, that makes sense! 

As for the lib file - I wonder If we could take advantage of the `prependData` property in webpack for bundling certain css files? I've used it in the past and it's helped to at least not have to remember to add the import for variables everytime. Not sure if that would actually help performance though.

**Comment by maureenlholland** [comment]
> it's helped to at least not have to remember to add the import for variables everytime

I am on board with less remembering!

**Comment by nathan-barrett** [comment]
@alexgibson I would love the idea of moving away from more snowflake pages, but I wonder how it would work if you are using multiple protocol components? (e.g. split, and picto)

**Comment by reemhamz** [comment]
So just trying to wrap my head around this, we could create a specific CSS bundle holding protocol components frequently used, which we will then import to a page:

```
page.html
{% block page_css %}
 {{ css_bundle('protocol_components_bundle') }}
{% endblock %}
```

which means we don't necessarily need to import these into our CSS pages:
```
page.css
@import '~@mozilla-protocol/core/protocol/css/components/call-out';
@import '~@mozilla-protocol/core/protocol/css/components/card';
@import '~@mozilla-protocol/core/protocol/css/components/modal';
...
```

 as well as import macros into the HTML:
```
page.html
{% from "macros-protocol.html" import split, call_out, zap, card, call_out_compact, ... with context %}
```

Did I understand this correctly? If so, this sounds super neat and I'm on board for this change. I've found it quite cumbersome to refer to other CSS files for examples + to make sure I have the correct paths (especially when working with wordmarks and logos)

**Comment by alexgibson** [comment]
Rather than 

```jinja
{% block page_css %}
 {{ css_bundle('protocol_components_bundle') }}
{% endblock %}
```

A template CSS block would look similar to:

```jinja
{% block page_css %}
 {{ css_bundle('c-split') }}
 {{ css_bundle('c-card) }}
 {{ css_bundle('page-css') }}
{% endblock %}
```

You would still need to import macros in templates, yes.

**Comment by reemhamz** [comment]
Started small with this ticket: https://github.com/mozilla/bedrock/pull/11107. Happy to hear any thoughts or critique on bundle names, haha.

**Comment by reemhamz** [comment]
## Game plan

These are the components we'll create bundles for, since they have the highest number of CSS imports in our codebase: 
- split component: ~35 imports 
- card (coupled with card-layout): ~16 imports
- picto (coupled with multi-column): ~24 imports
- <del>logo (which holds logos and wordmarks): ~50+ imports</del> (will revisit later)
- callout: ~25 imports
- article: ~15 imports
- newsletter-form: ~17 imports
- emphasis-box: ~19 imports

I decided to make a PR for each component import-to-bundle conversion, this will make it easier for whoever reviewing to tackle the task, and not be so overwhelmed. 

Some PRs will have more than one component import-to-bundle conversion, since they might be residing in the same file, and it'll be quicker to update rather than wait on tackling it in another PR.

**Comment by alexgibson** [comment]
@reemhamz nice work here! It looks like the dependency list is all ticked off, are we good to close this one out?

**Comment by reemhamz** [comment]
@alexgibson yes! All good to close this epic :) Thanks for all the help and guidance here 🚀
